cluster/swarm/cluster.go

```diff
@@ -359,9 +359,13 @@ func (c *Cluster) removeEngine(addr string) bool {
 func (c *Cluster) monitorDiscovery(ch <-chan discovery.Entries, errCh <-chan error) {
 	// Watch changes on the discovery channel.
 	currentEntries := discovery.Entries{}
-	for {
+	for ch != nil || errCh != nil {
 		select {
-		case entries := <-ch:
+		case entries, ok := <-ch:
+			if !ok {
+				ch = nil
```

Comment:
@dongluochen 
(1) > The problem is this function doesn't own the 2 channels. They could be used (or will be used) in other places.
Yes, `monitorDiscovery` doesn't own the 2 channels. But I don't think it does matter:  

```
case entries, ok := <-ch:
        if !ok {
		ch = nil
                break
        }
```
If `ok` is `false`, this means the channel has been closed in other places. So it is not meaningful to still wait here to receive messages from this channel. The same logic should apply to `errCh` as well.

(2) > I don't see why setting them to `nil` has any benefit while posing a threat.
If don't set them to `nil`, is there any better method to prevent dead-loop of this `for` statement?

Thanks very much!